nifi-nar-bundles/nifi-mqtt-bundle/nifi-mqtt-processors/src/main/java/org/apache/nifi/processors/mqtt/SubscribeMQTT.java

```diff
@@ -0,0 +1,301 @@
+import org.eclipse.paho.client.mqttv3.MqttMessage;
+import org.eclipse.paho.client.mqttv3.IMqttDeliveryToken;
+import org.eclipse.paho.client.mqttv3.persist.MemoryPersistence;
+
+import java.util.Collections;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Properties;
+import java.util.Set;
+import java.util.concurrent.LinkedBlockingQueue;
+import java.io.OutputStream;
+import java.io.IOException;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.ScheduledThreadPoolExecutor;
+import java.util.concurrent.TimeUnit;
+
+
+@Tags({"subscribe", "MQTT", "IOT"})
+@CapabilityDescription("Subscribes to a topic and receives messages from an MQTT broker")
+@SeeAlso({PublishMQTT.class})
+@WritesAttributes({@WritesAttribute(attribute="broker", description="MQTT broker that was the message source"),
+    @WritesAttribute(attribute="topic", description="MQTT topic on which message was received"),
+    @WritesAttribute(attribute="qos", description="The quality of service for this message."),
+    @WritesAttribute(attribute="isDuplicate", description="Whether or not this message might be a duplicate of one which has already been received."),
+    @WritesAttribute(attribute="isRetained", description="Whether or not this message was from a current publisher, or was \"retained\" by the server as the last message published on the topic.")})
+public class SubscribeMQTT extends AbstractMQTTProcessor {
+
+    public static final PropertyDescriptor PROP_TOPIC_FILTER = new PropertyDescriptor.Builder()
+            .name("Topic Filter")
+            .description("The MQTT topic filter to designate the topics to subscribe to.")
+            .required(true)
+            .expressionLanguageSupported(false)
+            .addValidator(StandardValidators.NON_BLANK_VALIDATOR)
+            .build();
+
+    public static final PropertyDescriptor PROP_QOS = new PropertyDescriptor.Builder()
+            .name("Quality of Service(QoS)")
+            .description("The Quality of Service(QoS) to receive the message with. Accepts three values '0', '1' and '2'; '0' for 'at most once', '1' for 'at least once', '2' for 'exactly once'.")
+            .required(true)
+            .allowableValues("0", "1", "2")
+            .build();
+
+
+    public static final PropertyDescriptor PROP_MAX_QUEUE_SIZE = new PropertyDescriptor.Builder()
+            .name("Max Queue Size")
+            .description("The MQTT messages are always being sent to subscribers on a topic. If the 'Run Schedule' is significantly behind the rate at which the messages are arriving to this " +
+                    "processor then a back up can occur. This property specifies the maximum number of messages this processor will hold in memory at one time.")
+            .required(true)
+            .addValidator(StandardValidators.NON_NEGATIVE_INTEGER_VALIDATOR)
+            .build();
+
+    private String broker;
```

Comment:
This is more of a general comment. Please review all of these attributes to see if any of them can be modified by a different threads. If so, then make them ```volatile```.